src/Microsoft.Extensions.OData.Migration/QueryNodeTranslator.cs

```diff
@@ -0,0 +1,364 @@
+        /// </summary>
+        /// <param name="nodeIn">V3 AllNode</param>
+        /// <returns>V4 AllNode</returns>
+        public override QueryNode Visit(Data.OData.Query.SemanticAst.AllNode nodeIn)
+        {
+            IEnumerable<RangeVariable> translated = nodeIn.RangeVariables.Select(r => TranslateRangeVariable(r));
+            AllNode ret = new AllNode(new Collection<RangeVariable>(translated.ToList()), TranslateRangeVariable(nodeIn.CurrentRangeVariable));
+            ret.Source = Visit((dynamic)nodeIn.Source) as CollectionNode;
+            ret.Body = Visit((dynamic)nodeIn.Body) as SingleValueNode;
+            return ret;
+        }
+
+        /// <summary>
+        /// Translates a V3 AnyNode to V4 AnyNode
+        /// </summary>
+        /// <param name="nodeIn">V3 AnyNode</param>
+        /// <returns>V4 AnyNode</returns>
+        public override QueryNode Visit(Data.OData.Query.SemanticAst.AnyNode nodeIn)
+        {
+            List<RangeVariable> translated = nodeIn.RangeVariables.Select(r => TranslateRangeVariable(r)).ToList();
+            AnyNode ret = new AnyNode(new Collection<RangeVariable>(translated), TranslateRangeVariable(nodeIn.CurrentRangeVariable));
+            ret.Source = Visit((dynamic)nodeIn.Source) as CollectionNode;
+            ret.Body = Visit((dynamic)nodeIn.Body) as SingleValueNode;
+            return ret;
+        }
+
+        /// <summary>
+        /// Translates V3 BinaryOperatorNode to V4 BinaryOperatorNode
+        /// </summary>
+        /// <param name="nodeIn">V3 BinaryOperatorNode</param>
+        /// <returns>V4 BinaryOperatorNode</returns>
+        public override QueryNode Visit(Data.OData.Query.SemanticAst.BinaryOperatorNode nodeIn)
+        {
+            BinaryOperatorKind kind = (BinaryOperatorKind)(int)nodeIn.OperatorKind;
+            return new BinaryOperatorNode(kind, Visit((dynamic)nodeIn.Left) as SingleValueNode, Visit((dynamic)nodeIn.Right) as SingleValueNode);
+        }
+
+
+        /// <summary>
+        /// Translates V3 CollectionNavigationNode to V4 CollectionNavigationNode
+        /// </summary>
+        /// <param name="nodeIn">V3 CollectionNavigationNode</param>
+        /// <returns>V4 CollectionNavigationNode</returns>
+        public override QueryNode Visit(Data.OData.Query.SemanticAst.CollectionNavigationNode nodeIn)
+        {
+            IEdmPathExpression pathExpr = null;
+            if (nodeIn.EntitySet != null)
+            {
+                IEdmNavigationSource navigationSource = v4model.FindDeclaredNavigationSource(nodeIn.EntitySet.Name);
+                pathExpr = navigationSource.Path;
+            }
+            IEdmStructuredType v4Type = v4model.FindType(nodeIn.NavigationProperty.DeclaringType.GetFullTypeName()) as IEdmStructuredType;
+            IEdmNavigationProperty v4navProperty = v4Type.FindProperty(nodeIn.NavigationProperty.Name) as IEdmNavigationProperty;

```

Comment:
What if the property cannot be found? Is it an error condition or can it be ignored?